src/server/asyncrpcprovider.ts

```diff
@@ -0,0 +1,44 @@
+export interface RpcRequest {
+	jsonrpc: '2.0';
+	method: string;
+	params?: any[];
+	id?: number;
+}
+
+interface RpcResponseBase {
+	jsonrpc: string;
+	id: number;
+}
+
+interface RpcResponseSuccess extends RpcResponseBase {
+	result: any;
+}
+
+interface RpcResponseFailed extends RpcResponseBase {
+	error: { code: number, message: string, data?: any };
+}
+
+type RpcResponse = RpcResponseSuccess | RpcResponseFailed;
+
+export class AsyncRpcProvider {
+	protected host;
+
+	constructor(host: any) {
+		this.host = host;
+	}
+	async call(request: RpcRequest): Promise<RpcResponse> {
+		const { id, method, params } = request || {};
+		if (!this.host.hasOwnProperty(method))
+			return { jsonrpc: '2.0', id: id!, error: { code: 0, message: 'unknown method' } };
```

Comment:
`id` is optional, only specified is a response is expected. I just didn't want TS to complain. We can probably just call the method and discard the result if `id` is undefined.

Reply:
looks like `id` is required in `RpcResponseBase` but we can make it optional for sure. 

And then you wouldn't need to put in the `!` since TS will know it's optional (and we can just shorthand it like 
```typescript
{ jsonrpc: '2.0', id, result };
```

Reply:
It's required in the response but not the request. We should just not send anything back if the `id` is missing.